include/range/v3/view/all.hpp

```diff
@@ -29,72 +31,89 @@ namespace ranges
     {
         /// \addtogroup group-views
         /// @{
+        template<typename Rng>
+        struct ref_view
+          : view_interface<ref_view<Rng>, range_cardinality<Rng>::value>
+        {
+        private:
+            CONCEPT_ASSERT(std::is_object<Rng>::value && Range<Rng>() && !View<Rng>());
+
+            Rng *rng_ = nullptr;
+        public:
+            ref_view() = default;
+            constexpr explicit ref_view(Rng &rng) noexcept
+              : rng_{&rng}
+            {}
+            constexpr iterator_t<Rng> begin() const
+            RANGES_AUTO_RETURN_NOEXCEPT
+            (
+                ranges::begin(*rng_)
+            )
+            constexpr sentinel_t<Rng> end() const
+            RANGES_AUTO_RETURN_NOEXCEPT
+            (
+                ranges::end(*rng_)
+            )
+            template<class R = Rng>
+            constexpr auto size() const
+                RANGES_DECLTYPE_NOEXCEPT(ranges::size(std::declval<R &>()))
+            {
+                return ranges::size(*rng_);
+            }
+            template<class R = Rng>
+            constexpr auto empty() const
+                RANGES_DECLTYPE_NOEXCEPT(ranges::empty(std::declval<R &>()))
+            {
+                return ranges::empty(*rng_);
+            }
+            template<class R = Rng>
+            constexpr auto data() const
+                RANGES_DECLTYPE_NOEXCEPT(ranges::data(std::declval<R &>()))
+            {
+                return ranges::data(*rng_);
+            }
+        };
+
```

Comment:
These members need constraints, per you recent discovery about the ills of SFINAE return types, no? It is that only for alias templates?

Reply:
I really want compilers to cache the result of overload resolution even when that result is failure, but for all I know this return type SFINAE could perform as badly as the alias template substitution failure case from #758. I'll constrain these.

Reply:
So constrained. I'm not married to the name `CanEmptyRange`.